internal/adapters/dataproviders/bitcoin/mempool_space/api.go

```diff
@@ -0,0 +1,516 @@
+	if err != nil {
+		return blockchain.BtcCoinbaseTransactionInformation{}, err
+	}
+
+	return blockchain.BtcCoinbaseTransactionInformation{
+		BtcTxSerialized:      serializedCoinbase.Bytes(),
+		BlockHash:            bitcoin.ToSwappedBytes32(block.Hash()),
+		BlockHeight:          blockInfo.Height,
+		SerializedPmt:        pmt,
+		WitnessMerkleRoot:    bitcoin.ToSwappedBytes32(merkle.CalcMerkleRoot(block.Transactions(), true)),
+		WitnessReservedValue: witnessReservedValue,
+	}, nil
+}
+
+func (api *MempoolSpaceApi) NetworkName() string {
+	return strings.ToLower(api.config.Name)
+}
+
+func (api *MempoolSpaceApi) GetBlockchainInfo() (blockchain.BitcoinBlockchainInfo, error) {
+	if err := api.validateNetwork(); err != nil {
+		return blockchain.BitcoinBlockchainInfo{}, err
+	}
+	const getBlockchainInfoError = "error getting blockchain info: %w"
+
+	blocks, err := api.GetHeight()
+	if err != nil {
+		return blockchain.BitcoinBlockchainInfo{}, fmt.Errorf(getBlockchainInfoError, err)
+	}
+
+	req, err := http.NewRequestWithContext(context.Background(), http.MethodGet, api.url+"/blocks/tip/hash", nil)
+	if err != nil {
+		return blockchain.BitcoinBlockchainInfo{}, fmt.Errorf(getBlockchainInfoError, err)
+	}
+	res, err := api.client.Do(req)
+	defer closeBody(res)
+	if err != nil {
+		return blockchain.BitcoinBlockchainInfo{}, fmt.Errorf(getBlockchainInfoError, err)
+	} else if err = handleErrorResponse(res); err != nil {
+		return blockchain.BitcoinBlockchainInfo{}, fmt.Errorf(getBlockchainInfoError, err)
+	}
+	data, err := io.ReadAll(res.Body)
+	if err != nil {
+		return blockchain.BitcoinBlockchainInfo{}, fmt.Errorf(getBlockchainInfoError, err)
+	}
+
+	return blockchain.BitcoinBlockchainInfo{
+		NetworkName:      api.NetworkName(),
+		ValidatedBlocks:  blocks,
+		ValidatedHeaders: blocks,
+		BestBlockHash:    string(data),
+	}, nil
+}
+
+func (api *MempoolSpaceApi) getBlock(blockHash string) (*btcutil.Block, error) {
+	req, err := http.NewRequestWithContext(context.Background(), http.MethodGet, fmt.Sprintf("%s/block/%s/raw", api.url, blockHash), nil)
+	if err != nil {
+		return nil, err
+	}
+	res, err := api.client.Do(req)
+	defer closeBody(res)
+	if err != nil {
+		return nil, err
+	} else if err = handleErrorResponse(res); err != nil {
+		return nil, err
+	}
+	data, err := io.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return btcutil.NewBlockFromBytes(data)
+}
+
+func (api *MempoolSpaceApi) validateNetwork() error {
+	supportedNetworks := []*chaincfg.Params{&chaincfg.TestNet3Params, &chaincfg.MainNetParams}
+	for _, network := range supportedNetworks {
+		if api.config.Name == network.Name {
+			return nil
+		}
+	}
+	return errors.New("unsupported network")
+}
+
+func closeBody(res *http.Response) {
+	if res == nil || res.Body == nil {
+		return
+	}
+	if err := res.Body.Close(); err != nil {
+		log.Error("Error closing body in MempoolSpace API:", err)
+	}
+}
+
+func handleErrorResponse(res *http.Response) error {
```

Comment:
Probably not the best name for this function. This function is handling not only the error response. If a response is a 2xx it will handle it by returning nil. What I mean is that no filtering has been made before executing this function to say that what will be handled here are only error responses.